src/wallet/rpcnames.cpp

```diff
@@ -354,10 +355,286 @@ saltMatchesHash(const valtype& name, const valtype& rand, const valtype& expecte
     return (Hash160(toHash) == uint160(expectedHash));
 }
 
+bool existsName(const valtype& name, const ChainstateManager& chainman)
+{
+    LOCK(cs_main);
+    const auto& coinsTip = chainman.ActiveChainstate().CoinsTip();
+    CNameData oldData;
+    return (coinsTip.GetName(name, oldData) && !oldData.isExpired());
+}
+
+bool existsName(const std::string& name, const ChainstateManager& chainman)
+{
+    return existsName(valtype(name.begin(), name.end()), chainman);
+}
+
 } // anonymous namespace
 
 /* ************************************************************************** */
 
+RPCHelpMan
+name_autoregister()
+{
+    NameOptionsHelp optHelp;
+    optHelp
+        .withNameEncoding()
+        .withWriteOptions()
+        .withArg("allowExisting", RPCArg::Type::BOOL, "false",
+                 "If set, then the name_new is sent even if the name exists already")
+        .withArg("delegate", RPCArg::Type::BOOL, "false",
+                 "If set, register a dd/ or idd/ name and delegate the name there. Name must be in d/ or id/ namespace.");
+
+    return RPCHelpMan("name_autoregister",
+        "\nAutomatically registers the given name; performs the first half and queues the second."
+            + HELP_REQUIRING_PASSPHRASE,
+        {
+            {"name", RPCArg::Type::STR, RPCArg::Optional::NO, "The name to register"},
+            {"value", RPCArg::Type::STR, RPCArg::Optional::OMITTED_NAMED_ARG, "Value for the name"},
+            optHelp.buildRpcArg(),
+        },
+        RPCResult {RPCResult::Type::ARR_FIXED, "", "",
+            {
+                {RPCResult::Type::STR_HEX, "txid", "the txid, used in name_firstupdate"},
+                {RPCResult::Type::STR_HEX, "rand", "random value, as in name_firstupdate"},
+            },
+        },
+        RPCExamples {
+            HelpExampleCli("name_autoregister", "\"myname\"")
+            + HelpExampleRpc("name_autoregister", "\"myname\"")
+        },
+        [&] (const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    if (!wallet)
+        return NullUniValue;
+    CWallet* const pwallet = wallet.get();
+
+    RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VSTR, UniValue::VOBJ});
+    const auto& chainman = EnsureAnyChainman(request.context);
+
+    // build a name_new
+    // broadcast it
+    // build a name_fu with nSequence
+    // queue it
+    // return the txid and rand from name_new, as an array
+    //
+    // if delegate=true:
+    //   build nn1 = d/something
+    //   build nn2 = dd/something (if already exist, something else)
+    //   broadcast nn1,nn2
+    //   build nfu1 = "d/something points to dd/xxx"
+    //   build nfu2 = "d/xxx is the value"
+    //   queue nfu1,nfu2
+    //   return txid from nn1
```

Comment:
Is that delegation stuff something we really want to do in Namecoin Core itself?  So far, we've managed to keep Namecoin Core agnostic to any kind of values or names, and handle all namespace and name-value-JSON stuff on a higher level.  I'd prefer to keep it that way.

Reply:
That's a fair point. I only added it because there was funding to do it. I agree with you there's something slightly ungainly about Namecoin Core going this far, and in particular about generating the JSON.

I'm not wholesale opposed to having it in, because it's better to have one good implementation in Namecoin Core than twelve thousand shoddy ones that never get maintained in various downstream consumers, but I think such functionality (that has to be aware of namespace rules etc) should at the very least be quite cleanly split off, like the proposal #365. (For delegation, having consumers issue two autoregisters and moving that further downstream honestly seems fine)

Maybe it could be a separate RPC, or behind a #ifdef directive?

Reply:
As you say, for delegation it is completely irrelevant from the consumer's side, and that's what it is doing for now, so I would just leave that to the caller.  (And if you use it from a "name registration" Qt UI, then the JSON generation should be in the Qt code, not the RPC one.)

Are you planning to add more complex JSON generation in the future?  In that case, I believe perhaps a separate library (in Python or Node or whatever typical consumers will use) that can be hooked onto the RPC interface makes the most sense.

Reply:
> Is that delegation stuff something we really want to do in Namecoin Core itself? So far, we've managed to keep Namecoin Core agnostic to any kind of values or names, and handle all namespace and name-value-JSON stuff on a higher level. I'd prefer to keep it that way.

@domob1812 I strongly disagree with this.  From a UX perspective, it is important for the Namecoin-Qt GUI to handle this, without expecting the user to go into a separate GUI.  (E.g. Tor Project policy is that user perception of the Tor/Firefox segregation should be minimized, because that segregation confuses users.)  In theory there could be a separate daemon that handle this, and accepts commands from Namecoin-Qt in a way that the user doesn't perceive, but the main advantage of such a thing is security/isolation, and I am not convinced that this is a case where that's needed.  (In fact, this functionality is so simple compared to the needed IPC logic that it may actually expose *more* attack surface to keep this in a separate process.)

Putting this functionality in the Qt code rather than RPC has the primary effect of making integration tests harder, so I am not convinced that that is a good idea.  (AIUI, Bitcoin Core policy is that functionality must be added to RPC before being added to Qt, so that integration testing is easier.)

Reply:
> > Is that delegation stuff something we really want to do in Namecoin Core itself? So far, we've managed to keep Namecoin Core agnostic to any kind of values or names, and handle all namespace and name-value-JSON stuff on a higher level. I'd prefer to keep it that way.
> 
> @domob1812 I strongly disagree with this. From a UX perspective, it is important for the Namecoin-Qt GUI to handle this, without expecting the user to go into a separate GUI. (E.g. Tor Project policy is that user perception of the Tor/Firefox segregation should be minimized, because that segregation confuses users.) In theory there could be a separate daemon that handle this, and accepts commands from Namecoin-Qt in a way that the user doesn't perceive, but the main advantage of such a thing is security/isolation, and I am not convinced that this is a case where that's needed. (In fact, this functionality is so simple compared to the needed IPC logic that it may actually expose _more_ attack surface to keep this in a separate process.)
> 
> Putting this functionality in the Qt code rather than RPC has the primary effect of making integration tests harder, so I am not convinced that that is a good idea. (AIUI, Bitcoin Core policy is that functionality must be added to RPC before being added to Qt, so that integration testing is easier.)

I agree with @domob1812 I don't want Namecoin Core to be full of pre-defined namespaces. Namecoin Core should be free from d/  dd/ id/ idd/ and should not show .bit like @JeremyRand want it. For over 10 years Namecoin was perfectly fine, now we not going change that. Namespaces should be left untouched.